### ISSUE TITLE
Python temp atomspace usage

### ISSUE THREAD

**Issue by linas** [opencog/atomspace#2410]
I believe that the python code, after pull rq #2498 no longer respects the temp atomspace.  Specifically, consider the following pseudocode:
```
(Get  (Variable "x")
   (And
       (Present (ListLink (Concept "foo") (Variable "x")))
       (Evaluation (GroundedPredicate "py:test_one")(List (Variable "x")))
       (Evaluation (GroundedPredicate "py:test_two")(List (Variable "x")))
    ))
```
with 
```
def test_one(v):
   thing = add_node(types.Concept, "barleycorn")
   thang = add_link(types.ListNode, thing, v, TruthValue(0.9, 0.75))
   thong = space_server.get(thang)
   return thong.getTruthValue()

def test_two(v):
    return TruthValue (0,1)   # false
```

In this example, the atoms `thing` `thang` and `thong` are supposed to be created in the temp atomspace, not the main atomspace.  The main atomspace is never polluted with these extra atoms. Doubly so, since the second test always fails, so the GetLink should always return the empty set.

That's the intent of the temp atomspace: to avoid polluting the main atomspace with any temporary atoms created during the search.

Before the pull req #2398, the `applier.apply_tv(as, schema.substr(pos), args);` API told python how to find the temp atomspace that will hold `thing`, `thang`, `thong`. Since that API is now gone, its not clear how this can work.

**Comment by linas** [comment]
I think the pseudocode for a working version would be something like this:
```
TruthValuePtr apply_tv(AtomSpace*as, const std::string& func, Handle varargs)
{
    this->eval("from opencog.atomspace import AtomSpace");
    this->eval("from opencog.type_constructors import set_type_ctor_atomspace");
    this->eval("set_type_ctor_atomspace(AtomSpace(" + std::to_string((uint64_t) as) + "))\n");

   etc...
}
```
and likewise for `apply()`

**Comment by linas** [comment]
@noskill @vsbogd follow-on to pull req #2398

**Comment by vsbogd** [comment]
Test `https://github.com/opencog/atomspace/blob/master/tests/scm/scm-python-shared-atomspace.scm` which tests this passes.
See https://circleci.com/gh/opencog/atomspace/4106?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-checks-link

Test log:
```
69/138 Testing: SCMPythonSharedAtomspace
69/138 Test: SCMPythonSharedAtomspace
Command: "/usr/bin/guile" "-L" "/ws/atomspace/build/opencog/scm" "--use-srfi=64" "/ws/atomspace/tests/scm/scm-python-shared-atomspace.scm"
Directory: /ws/atomspace/build/tests/scm
"SCMPythonSharedAtomspace" start time: Dec 02 13:04 UTC
Output:
----------------------------------------------------------
%%%% Starting test python-guile-shared-atomspace-test  (Writing full log to "python-guile-shared-atomspace-test.log")
# of expected passes      2
%%%% Exiting test suite
;;; note: auto-compilation is enabled, set GUILE_AUTO_COMPILE=0
;;;       or pass the --no-auto-compile argument to disable.
;;; compiling /ws/atomspace/tests/scm/scm-python-shared-atomspace.scm
;;; compiled /root/.cache/guile/ccache/2.2-LE-8-3.A/ws/atomspace/tests/scm/scm-python-shared-atomspace.scm.go
<end of output>
Test time =   0.22 sec
----------------------------------------------------------
Test Passed.
"SCMPythonSharedAtomspace" end time: Dec 02 13:04 UTC
"SCMPythonSharedAtomspace" time elapsed: 00:00:00
----------------------------------------------------------
``

**Comment by noskill** [comment]
Could you please provide example without space server?

I use python context manager to use temporary atomspaces:

```python
with tmp_atomspace() as tmp:
       execute_atom(tmp, atom)
```

**Comment by noskill** [comment]
I think python biding works like that for some time now, like 2 years or so. 
PR https://github.com/opencog/atomspace/pull/2398 removed only unused code.

**Comment by noskill** [comment]
@vsbogd perhaps Linas meant that during pattern matching new nodes should not be added to the atomspace if not matched by evaluation link..

**Comment by vsbogd** [comment]
Thanks @noskill, yes I probably didn't get the description at the first reading. So there is hypothesis that when PatternMatcher tries evaluating `GroundedPredicate` which in turn calls Python then all atoms created in Python are created in the main atomspace.

I suppose it is true and probably it was true before #2398. I am reopening this but we should write an unit test to find the problem.

**Comment by vsbogd** [comment]
BTW I don't think we need space server or Sheme to check this it should be reproducible in pure Python. Just BindLink with Evaluation which creates atoms inside.

**Comment by linas** [comment]
Re: spaceserver: that code was pseudocode, it was meant to be a stand-in for "some complicated blackbox that creates Atoms and does other computations."

The implementation of "tmp_atomspace()" is clearly wrong, in two different ways: first, its not actually using the supplied temp atomspace, its just creating a new one.  Second, atomspace initializing takes a fair amount of CPU time -- its slow. You don't want to do this too often. That's why the pattern matcher keeps a pool of 8 different pre-initialized temp atomspaces, ready to go -- to avoid the overhead.

So for example, the same `Get` as before, but this time:
```
def test_one(v):
    # Add atom to temp atomspace
   thing = add_node(types.Concept, "barleycorn", TruthValue(0.9, 0.75))
   return TruthValue(1, 1)

def test_two(v):
    # Get Atom from temp atomspace; assume its the SAME one!
    # This can't be called with a different atomspace (that doesn't contain `thing`)
    return thing.getTruthValue()   
```

**Comment by linas** [comment]
Also, grepping for `tmp_stomspace` shows that it is not used anywhere. I think it can be safely removed. I think its a left-over from .. a bygone era.

**Comment by noskill** [comment]
tmp_atomspace is used in https://github.com/singnet/semantic-vision

**Comment by noskill** [comment]
This example may be rewritten

```
def test_one(v):
    # Add atom to temp atomspace
   atomspace = v.atomspace
   thing = atomspace.add_node(types.Concept, "barleycorn", TruthValue(0.9, 0.75))
   return TruthValue(1, 1)
```
tmp atomspace created by pattern matcher is used here to create atoms

**Comment by noskill** [comment]
Having said that i think that setting atomspace for atom constructors should done on c++ side.
Wrapping atomspace in pyobject then monkey-patching module with it, then unwrapping pyobject are unnecessary steps. I think instead ExecutionOutputLink should set thread local pointer to atomspace.

**Comment by linas** [comment]
Closing, this looks like its fixed by #2420  Thanks!
